CHANGELOG.md

```diff
@@ -5,6 +5,13 @@
 * [CHANGE] Querier / ruler: Change `-querier.max-fetched-chunks-per-query` configuration to limit to maximum number of chunks that can be fetched in a single query. The number of chunks fetched by ingesters AND long-term storare combined should not exceed the value configured on `-querier.max-fetched-chunks-per-query`. #4260
 * [ENHANCEMENT] Add timeout for waiting on compactor to become ACTIVE in the ring. #4262
 * [ENHANCEMENT] Reduce memory used by streaming queries, particularly in ruler. #4341
+* [ENHANCEMENT] Ring: allow experimental configuration of disabling of heartbeat timeouts by setting the relevant configuration value to zero. Applies to the following: #4342
+  * `-distributor.ring.heartbeat-timeout`
+  * `-ring.heartbeat-timeout`
+  * `-ruler.ring.heartbeat-timeout`
+  * `-alertmanager.sharding-ring.heartbeat-timeout`
+  * `-compactor.ring.heartbeat-timeout`
+  * `-store-gateway.sharding-ring.heartbeat-timeout`
```

Comment:
I debated whether we should only allow this on the ingester (the intended use case). The alternative to this would be to add validation to the other configuration options to disallow 0.

Reply:
The workload resharding in all services except ingesters (so, compactor, ruler, alertmanager, store-gateway) is based on the heartbeat. If we disable it, the workload assigned to an unhealthy instance will be never resharded to an healthy one. Because of this, I'm a bit dubious if we should allow disabling the heartbeat timeout for other services, until we'll have a viable alternative to reshard the workload.

Reply:
Agreed, it's not necessary. I'll add the validation instead.

Reply:
I was checking how the resharding work on unhealthy instances of other services (ruler, compactor, alertmanager, store-gateway). They all work the same way. If an instance is unhealthy in the ring, its workload it not resharded to other healthy replicas, so allowing to disable the heartbeat for these services should basically have no real effect on this regards.

Reply:
Because of this, I don't see any problem allowing to disable heartbeat in all these services too.

Reply:
> Agreed, it's not necessary. I'll add the validation instead.

I deleted a previous erroneous comment. Please see new ones.

Reply:
OK - will leave as-is.